### PR TITLE
[CPU] ROIAlign: dynamic shapes support

### PR DESCRIPTION
**Pull request by v-Golubev**
### Details:
 - *ROIAlign: dynamic shapes support*
 - *ROIAlign: added CPU tests with dynamic shapes*

### Tickets:
 - *70688*

### CODE REVIEW

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp [hunk 1]
```diff
@@ -3,108 +3,172 @@
...
+            for (size_t i = 0; i < coordsTensor.get_size(); i += 4) {
+                coordsTensorData[i] = 1.f;
+                coordsTensorData[i] = 1.f;
+                coordsTensorData[i] = 19.f;
+                coordsTensorData[i] = 19.f;
+            }
+        } else if (coordsET == ElementType::bf16) {
+            auto coordsTensorData = static_cast<std::int16_t*>(coordsTensor.data());
+            for (size_t i = 0; i < coordsTensor.get_size(); i += 4) {
+                coordsTensorData[i] = static_cast<std::int16_t>(ngraph::float16(1.f).to_bits());
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp:107]: as I can see we used some different approach to generate `bf16` values.
https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/ie_test_utils/common_test_utils/data_utils.hpp#L259

where did you take this approach?

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp [hunk 2]
```diff
@@ -3,108 +3,172 @@
...
+        function = std::make_shared<ngraph::Function>(results, ngraph::ParameterVector{ float_params[0], float_params[1], int_params[0] }, "ROIAlign");
     }
 };
 
 TEST_P(ROIAlignLayerCPUTest, CompareWithRefs) {
     SKIP_IF_CURRENT_TEST_IS_DISABLED()
-    Run();
-    CheckPluginRelatedResults(executableNetwork, "ROIAlign");
+    run();
+    //CheckPluginRelatedResults(executableNetwork, "ROIAlign");
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp:171]: please don't forget uncomment

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp [hunk 3]
```diff
@@ -3,108 +3,172 @@
...
-        selectedType = std::string("unknown_") + inPrc.name();
+        selectedType = std::string("unknown_") + InferenceEngine::details::convertPrecision(netPrecision).name();
 
-        threshold = 1e-2;
+        if (netPrecision == ElementType::bf16) {
+            rel_threshold = 1e-2;
+        }
         const ngraph::ResultVector results{std::make_shared<ngraph::opset3::Result>(roialign)};
-        function = std::make_shared<ngraph::Function>(results, params, "ROIAlign");
+        function = std::make_shared<ngraph::Function>(results, ngraph::ParameterVector{ float_params[0], float_params[1], int_params[0] }, "ROIAlign");
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp:164]: use `makeNgraphFunction` instead manually initialization `roialign->get_rt_info() = getCPUInfo();`

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp [hunk 4]
```diff
@@ -93,15 +89,15 @@ void MKLDNNROIAlignNode::getSupportedDescriptors() {
         IE_THROW() << errorPrefix << "doesn't support output with rank: " << getOutputShapeAtPort(0).getRank();
     }
 
-    if (getInputShapeAtPort(1).getStaticDims()[1] != 4) {
-        IE_THROW() << errorPrefix << "has invalid shape on 1st input: ["
-                           << getInputShapeAtPort(1).getStaticDims()[0] << "," << getInputShapeAtPort(1).getStaticDims()[1] << "]";
+    const auto& proposalsDims = getInputShapeAtPort(1).getDims();
+    if (proposalsDims[1] != Shape::UNDEFINED_DIM && proposalsDims[1] != 4) {
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp]: Corrected, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp [hunk 5]
```diff
@@ -118,7 +114,7 @@ void MKLDNNROIAlignNode::initSupportedPrimitiveDescriptors() {
     }
 
     NodeConfig config;
-    config.dynBatchSupport = false;
+    config.dynBatchSupport = true;
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp]: The change was reverted, thanks

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp [hunk 6]
```diff
@@ -3,108 +3,172 @@
...
-        result << std::to_string(obj.index);
+
+        result << netPrecision << "_IS=";
+        for (const auto& shape : inputShapes) {
+            result << CommonTestUtils::partialShape2str({ shape.first }) << "_";
+        }
+        result << "TS=";
+        for (const auto& shape : inputShapes) {
+            result << "(";
+            if (!shape.second.empty()) {
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp]: Yes, we can. Corrected, thanks

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp (hunk 1 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp]: BF16 data generations was fixed, thank you!

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp (hunk 2 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp]: Done, thanks

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp (hunk 3 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/roialign.cpp]: Fixed, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp [hunk 7]
```diff
@@ -93,15 +89,15 @@ void MKLDNNROIAlignNode::getSupportedDescriptors() {
...
-        IE_THROW() << errorPrefix << "has invalid shape on 1st input: ["
-                           << getInputShapeAtPort(1).getStaticDims()[0] << "," << getInputShapeAtPort(1).getStaticDims()[1] << "]";
+    const auto& proposalsDims = getInputShapeAtPort(1).getDims();
+    if (proposalsDims[1] != Shape::UNDEFINED_DIM && proposalsDims[1] != 4) {
+        IE_THROW() << errorPrefix << "has invalid shape on 1st input: [" << proposalsDims[0] << "," << proposalsDims[1] << "]";
     }
 
-    if (getInputShapeAtPort(1).getStaticDims()[0] != getInputShapeAtPort(2).getStaticDims()[0]) {
+    const auto& indexesDims = getInputShapeAtPort(2).getDims();
+    if (proposalsDims[0] != indexesDims[0]) {
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp]: Done, thank you

**Review by mandrono — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@v-Golubev please make CI green

**Pull request merged**